ably/realtime_channel_integration_test.go

```diff
@@ -93,6 +93,88 @@ func TestRealtimeChannel_Subscribe(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestRealtimeChannel_SubscriptionFilters(t *testing.T) {
+	app, err := ablytest.NewSandbox(nil)
+	assert.NoError(t, err)
+	defer app.Close()
+	options := app.Options()
+	restClient, err := ably.NewREST(options...)
+	assert.NoError(t, err)
+	msg := []*ably.Message{
+		{
+			Name: "filtered",
+			Data: "This message will be filtered",
+			Extras: map[string]interface{}{
+				"headers": map[string]interface{}{
+					"name":   "value one",
+					"number": 1234,
+					"bool":   true,
+				},
+			},
+		},
+		{
+			Name: "filtered",
+			Data: "filtered messages",
+		},
+		{
+			Name: "filtered",
+			Data: "This message will be filtered because it does not meet condition on headers.number",
+			Extras: map[string]interface{}{
+				"headers": map[string]interface{}{
+					"name":   "value one",
+					"number": 5678,
+					"bool":   true,
+				},
+			},
+		},
+	}
+	filter := ably.DeriveOptions{
+		Filter: "name == `\"filtered\"` && headers.number == `1234`",
+	}
+
+	realtimeClient := app.NewRealtime(ably.WithEchoMessages(false))
+	defer safeclose(t, ablytest.FullRealtimeCloser(realtimeClient))
+
+	err = ablytest.Wait(ablytest.ConnWaiter(realtimeClient, realtimeClient.Connect, ably.ConnectionEventConnected), nil)
+	assert.NoError(t, err)
+
+	restChannel := restClient.Channels.Get("test")
+	rtDerivedChannel, err := realtimeClient.Channels.GetDerived("[?param=1]test", filter)
+	realtimeChannel := realtimeClient.Channels.Get("test")
+	assert.NoError(t, err, "realtimeChannel: GetDerived()=%v", err)
+
+	err = realtimeChannel.Attach(context.Background())
+	assert.NoError(t, err,
+		"realtimeClient: Attach()=%v", err)
+
+	rtDerivedSub, unsub, err := ablytest.ReceiveMessages(rtDerivedChannel, "")
+	assert.NoError(t, err, "realtimeClient:.Subscribe(context.Background())=%v", err)
+	defer unsub()
+	rtSub, unsub2, err := ablytest.ReceiveMessages(realtimeChannel, "filtered")
+	assert.NoError(t, err, "realtimeClient:.Subscribe(context.Background())=%v", err)
+	defer unsub2()
+
+	err = restChannel.PublishMultiple(context.Background(), msg)
+	assert.NoError(t, err, "restClient: PublishMultiple()=%v", err)
+
+	timeout := 15 * time.Second
+
+	// Check message received on rtDerivedChannel
+	err = expectMsg(rtDerivedSub, "filtered", msg[0].Data, timeout, true)
+	assert.NoError(t, err)
+	// Ensure that no additional message is received on rtDerivedChannel
+	err = expectMsg(rtDerivedSub, "filtered", "incorrect data", timeout, true)
+	assert.Error(t, err)
+	assert.Equal(t, err, fmt.Errorf("waiting for message name=%q data=%q timed out after %v", "filtered", "incorrect data", timeout))
+	// Check messages received on realtimeChannel
+	err = expectMsg(rtSub, "filtered", msg[0].Data, timeout, true)
```

Comment:
Rather than waiting 15s for this test to pass, could we use an "end" message like we do for the js tests?